File: tree_constructor/platform_adapter_go/request/send_req.go
package request

import (
	"fmt"
	"log"
	"rb/platform_adapter"
	"rb/platform_adapter/aws"
	"rb/platform_adapter/docker"
	"rb/platform_adapter/openlambda"
	"rb/util"
	"rb/workload"
	"strconv"
	"sync"
	"time"
)

type RunOptions struct {
	PlatformType string
	Workload     *workload.Workload
	StartOptions map[string]interface{}
	KillOptions  map[string]interface{}
	Config       interface{} // path to config file or map of config
	Tasks        int
	Timeout      int
	TotalTime    int
}

var seen = make(map[string]int)
var seenLock = &sync.Mutex{}

func getId(name string) string {
	seenLock.Lock()
	defer seenLock.Unlock()
	if _, ok := seen[name]; !ok {
		seen["id"] = 0
	}
	seen[name] += 1
	return name + "_" + strconv.Itoa(seen[name])
}

func deployFuncs(funcs []workload.Function, platform platform_adapter.PlatformAdapter) error {
	err := platform.DeployFuncs(funcs)
	return err
}

func task(platform platform_adapter.PlatformAdapter, timeout int, reqQ chan workload.Call, errQ chan error) {
	for {
		select {
		case req, ok := <-reqQ:
			if !ok {
				return
			}

			done := make(chan error, 1)
			go func() {
				options := make(map[string]interface{})
				options["invoke_id"] = getId(req.Name)
				options["req"] = util.GetCurrTime()
				err := platform.InvokeFunc(req.Name, timeout, options)
				done <- err
			}()

			select {
			case err := <-done:
				if err != nil {
					errQ <- fmt.Errorf("failed to invoke function %s: %s", req.Name, err)
				} else {
					errQ <- nil
				}
			}
		}
	}
}

func run(calls []workload.Call, tasks int, platform platform_adapter.PlatformAdapter, timeout int, totalTime int) (map[string]interface{}, error) {
	/*	workload: the workload to run
		num_tasks: the number of tasks to run
		platform: the platform to run on
		timeout: the timeout for each task
		total_time: the total time to run the workload
		returns: the number of tasks that were run
	*/
	reqQ := make(chan workload.Call, 64)
	errQ := make(chan error)
	for i := 0; i < tasks; i++ {
		go task(platform, timeout, reqQ, errQ)
	}
	t0 := time.Now()

	callIdx := 0
	waiting := 0
	fails := 0
	successes := 0
	progressSuccess := 0
	progressSnapshot := 0.0
	elapsed := 0.0
	start := time.Now()
	for calls != nil && len(calls) > 0 {
		// if time is up while totalTime is set,
		// or all tasks are done while totalTime is not set, break
		if (totalTime > 0 && elapsed > float64(totalTime)) ||
			((successes+fails+waiting) == len(calls) && totalTime <= 0) {
			break
		}

		select {
		case reqQ <- workload.Call{Name: calls[callIdx].Name}:
			waiting += 1
			callIdx = (callIdx + 1) % len(calls)
		case err := <-errQ:
			if err != nil {
				fails += 1
				fmt.Printf("%s\n", err.Error())
			} else {
				successes += 1
				progressSuccess += 1
			}
			waiting -= 1
		}

		elapsed = time.Since(start).Seconds()
		// show throughput stats about every 1 seconds
		if elapsed > progressSnapshot+1 {
			log.Printf("throughput: %.1f/second\n", float64(progressSuccess)/(elapsed-progressSnapshot))
			progressSnapshot = elapsed
			progressSuccess = 0
		}
	}
	// if there are still tasks running, wait for them to finish
	for waiting > 0 {
		err := <-errQ
		if err != nil {
			fails += 1
			fmt.Printf("%s\n", err.Error())
		} else {
			progressSuccess += 1
			successes += 1
		}
		waiting -= 1

		elapsed = time.Since(start).Seconds()
		// show throughput stats
		if elapsed > progressSnapshot+1 {
			log.Printf("throughput: %.1f/second\n", float64(progressSuccess)/(elapsed-progressSnapshot))
			progressSnapshot = elapsed
			progressSuccess = 0
		}
	}
	close(reqQ)

	t1 := time.Now()

	seconds := t1.Sub(t0).Seconds()
	throughput := float64(successes) / seconds

	return map[string]interface{}{
		"ops/s":   throughput,
		"seconds": seconds,
	}, nil
}

func newPlatformAdapter(platformType string) platform_adapter.PlatformAdapter {
	var platform platform_adapter.PlatformAdapter
	switch platformType {
	case "openlambda":
		platform, _ = openlambda.NewOpenLambda()
	case "docker":
		platform, _ = docker.NewDockerPlatform()
	case "aws":
		platform, _ = aws.NewAWSPlatform()
	default:
		platform = nil
	}
	return platform
}

// AutoRun start worker, deploy functions, run workload, kill worker
func AutoRun(opts RunOptions) (map[string]interface{}, error) {
	platform := newPlatformAdapter(opts.PlatformType)
	if platform == nil {
		return nil, fmt.Errorf("unsupported platform type: %s", opts.PlatformType)
	}

	err := platform.LoadConfig(opts.Config)
	if err != nil {
		return nil, err
	}

	if opts.PlatformType == "aws" {
		err = platform.StartWorker(opts.StartOptions)
		if err != nil {
			log.Fatalf("failed to start worker: %v", err)
		}

		if err := deployFuncs(opts.Workload.Funcs, platform); err != nil {
			log.Fatalf("failed to deploy functions: %v", err)
		}
	} else {
		if err := deployFuncs(opts.Workload.Funcs, platform); err != nil {
			log.Fatalf("failed to deploy functions: %v", err)
		}

		err = platform.StartWorker(opts.StartOptions)
		if err != nil {
			log.Fatalf("failed to start worker: %v", err)
		}
	}

	runStats, err := run(opts.Workload.Calls, opts.Tasks, platform, opts.Timeout, opts.TotalTime)
	if err != nil {
		platform.KillWorker(nil)
		log.Fatalf("failed to run workload: %v", err)
	}

	err = platform.KillWorker(opts.KillOptions)
	if err != nil {
		log.Fatalf("failed to kill worker: %v", err)
	}
	runStats = util.Union(runStats, platform.GetStats())

	return runStats, nil
}

File: tree_constructor/send_req.py
import os
import shutil
import random

import requests
import json
import threading
import time
import sys
import concurrent.futures
from concurrent.futures import ThreadPoolExecutor

seen = {}
seen_lock = threading.Lock()


def get_curr_time():
    return time.time() * 1000


def get_id(name):
    global seen
    with seen_lock:
        if name not in seen:
            seen[name] = 1
        id = str(seen[name])
        seen[name] += 1
        return name + "_" + id

# deploy concurrently
def deploy_funcs(workload, platform):
    funcs = workload["funcs"]
    call_set = {call["name"] for call in workload["calls"]}
    with ThreadPoolExecutor(max_workers=8) as executor:
        futures = []
        for fn in funcs:
            # only deploy functions that are called to save time
            if fn["name"] not in call_set:
                continue
            meta = fn["meta"]
            code = "\n".join(fn["code"])
            func_config = {
                "name": fn["name"],
                "code": code,
                "requirements_txt": meta["requirements_txt"]
            }
            future = executor.submit(platform.deploy_func, func_config)
            futures.append(future)
        for future in futures:
            future.result()
    return workload


def run(workload, num_tasks, platform, timeout=None, total_time=None):
    deploy_funcs(workload, platform)
    calls = workload['calls']
    finished = 0
    start_time = time.time()

    stamp = time.time()
    with ThreadPoolExecutor(max_workers=num_tasks) as executor:
        futures = [executor.submit(task, call, platform) for call in calls]
        for future in futures:
            try:
                if timeout is None:
                    future.result()
                else:
                    future.result(timeout=timeout)
            except concurrent.futures.TimeoutError:
                print("Task timed out")
            except Exception as e:
                print(f"Task resulted in an exception: {e}")
            finished += 1
            current_time = time.time()
            if finished % 1000 == 0:
                print("current throughput: ", 1000 / (current_time - stamp))
                stamp = time.time()

        # if after running the calls, there is still time remaining, then pick some calls randomly
        while total_time and time.time() - start_time < total_time:
            remaining_futures = [executor.submit(task, call, platform) for call in calls]
            for future in remaining_futures:
                try:
                    future.result()
                except concurrent.futures.TimeoutError:
                    print("Task timed out")
                except Exception as e:
                    print(f"Task resulted in an exception: {e}")

                finished += 1
                if finished % 1000 == 0:
                    print("current throughput: ", 1000 / (time.time() - stamp))
                    stamp = time.time()
                if time.time() - start_time >= total_time:
                    executor.shutdown(wait=False, cancel_futures=True)
                    break
    # clean up
    global seen
    with seen_lock:
        seen={}
    end_time = time.time()
    seconds = end_time - start_time
    return seconds, finished / seconds


def task(call, platform):
    req_body = {
        "invoke_id": get_id(call['name']),
        "req": get_curr_time()
    }
    options = {
        "req_body": req_body
    }
    resp_body,err = platform.invoke_func(call['name'], options=options)

    if resp_body is None or resp_body == "":
        raise Exception(f"Error: {err}")
